Jetpack Connect: Block blacklisted sites

tyxla opened the pull request:
#### Changes proposed in this Pull Request

* Introduce a blacklisted site notice and its corresponding notice type.
* Refine `checkUrl` action error to contain only what we might ever need from it.
* Store the JPC authorization errors in the state.
* Introduce a `isSiteBlacklistedError` selector.
* Block blacklisted sites in URL entry step. 
* Block blacklisted sites in auth step.

#### Preview

Blocking blacklisted sites in the site URL entry step:
![](https://cldup.com/6rXT6_bgrb.png)

Blocking blacklisted sites in the authorization step:
![](https://cldup.com/Xl6aCP-DZb.png)

#### Testing instructions

* Checkout this branch.
* If you haven't done that yet, follow instructions in the following diff: 
* Create a basic JN site from [this link](http://jurassic.ninja/create).
* Connect the site, selecting a free plan.
* Blacklist the site by adding the "Black" flag in the RC of the blog.
* Disconnect the site.
* Attempt to connect the site, starting from wp-admin (add `&calypso_env=development` to the connect button URL).
* Verify that in the auth step you can see the error as shown on the screenshot.
* Go to http://calypso.localhost:3000/jetpack/connect
* Input the site URL and submit the form.
* Verify you can see the error in the same page, before the auth process has started.
* Test both the Jetpack and the Calypso connection flows with a non-blacklisted site, verify they still work as expected.

This fixes 3155-gh-jpop-issues.

ockham commented:
> * Attempt to connect the site, starting from wp-admin (add &calypso_env=development to the connect button URL).
> * Verify that in the auth step you can see the error as shown on the screenshot.

Not really seeing any error on the first screen -- not sure that's expected?

![bildschirmfoto von 2019-01-08 18-00-54](https://user-images.githubusercontent.com/96308/50846563-c6da3e80-136f-11e9-9be7-831ee365011b.png)

Second screen shows an error :+1: 

![bildschirmfoto von 2019-01-08 18-01-33](https://user-images.githubusercontent.com/96308/50846582-cfcb1000-136f-11e9-8f51-7e684763f76b.png)

ockham commented:
> - Go to http://calypso.localhost:3000/jetpack/connect
> - Input the site URL and submit the form.
> - Verify you can see the error in the same page, before the auth process has started.

This isn't what I'm seeing -- instead, I'm taken to the plans selection step (`/jetpack/connect/plans/:site`). If I select Free, I end up at the onboarding checklist (`/plans/my-plan/:site?tour=jetpack`, stuck in placeholder state), however with the site connection marked as broken in the sidebar.

![site-connection-broken](https://user-images.githubusercontent.com/96308/50847160-2edd5480-1371-11e9-84bf-2450bc27174d.png)

tyxla commented:
> Not really seeing any error on the first screen -- which is probably expected.

Yeah, that's expected - the user needs to confirm their intent first before we're able to continue.

> Second screen shows an error 👍

Awesome!

> If I click 'Try again' (Erneut versuchen in the screenshot above), I'm redirected back to wp-admin, where I'm shown this error:
> 
> Should we be clearer here? And maybe also disable the 'Try again' button?

Yeah, it should be clear that this is intentionally a dead end for the user, so I've completely removed the "Try again" button for that case in https://github.com/Automattic/wp-calypso/pull/29983/commits/935a717c108bb2b18eb199ea02552fcb64bea6f1:

![](https://cldup.com/qV5XfTDouY.png)

> This isn't what I'm seeing -- instead, I'm taken to the plans selection step (`/jetpack/connect/plans/:site`). If I select Free, I end up at the onboarding checklist (`/plans/my-plan/:site?tour=jetpack`, stuck in placeholder state), however with the site connection marked as broken in the sidebar.

Hmm, are you sure you had this branch checked out, the related diff applied and the API sandboxed before testing? I'm unable to reproduce. You're not supposed to be able to reach the plans page - the error should be displayed at the site entry step, as shown on the screenshot.

Could I ask you to re-test? I'm also happy to pair on that. Thanks!

ockham commented on client/jetpack-connect/main.jsx:
```diff
@@ -243,11 +244,25 @@ export class JetpackConnectMain extends Component {
+			this.isCurrentUrlFetched() &&
+			this.props.jetpackConnectSite &&
+			this.props.jetpackConnectSite.error &&
+			this.props.jetpackConnectSite.error.error === error
```
Could use `_.get()` for the last three lines.

tyxla replied:
Good call 👍 done in https://github.com/Automattic/wp-calypso/pull/29983/commits/239806677baadbaae57deacb34638fbe75fb8e15.

ockham commented on client/state/jetpack-connect/test/selectors.js:
```diff
@@ -371,6 +372,56 @@ describe( 'selectors', () => {
+					},
+				},
+			},
+		};
```
Personally, I think I'd move each of these states into their corresponding `test()`s, but that's obviously a matter of taste.

tyxla replied:
👍 done in https://github.com/Automattic/wp-calypso/pull/29983/commits/46e6d3af82ef4520f8f680009a691f4a164357b7

ockham commented:
Tried again, and seeing the same behavior as before :slightly_frowning_face: 

- Trying to reconnect from wp-admin (after blacklisting and disconnecting), I'm not able to finish the process (great to disable the authorize button now BTW)

![image](https://user-images.githubusercontent.com/96308/51007396-90efb280-1547-11e9-98e2-54bfa6621a3e.png)

- Starting from http://calypso.localhost:3000/jetpack/connect/ however, I'm again taken through to the plans page.

> Hmm, are you sure you had this branch checked out, the related diff applied and the API sandboxed before testing?

Yeah, all of these. (I don't think I'd be seeing the message when starting from wp-admin otherwise, would I?)

I also tried the other way round, i.e. after blacklisting and disconnecting, I started at http://calypso.localhost:3000/jetpack/connect/ Once again, entering the site's URL and clicking connect took me to the plans page.

In case it matters, I'm doing this in the same incognito window where I keep the site's wp-admin open in a tab.

tyxla commented:
As discussed earlier via Slack, it seems that blacklisting before disconnecting leads to weird results. I've updated the test instructions, so disconnection goes first.

I've also addressed the feedback. Would appreciate another set of testing and review. Thanks!
